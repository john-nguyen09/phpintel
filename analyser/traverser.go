package analyser

import (
	"github.com/john-nguyen09/go-phpparser/phrase"
)

func Traverse(node phrase.AstNode, analyser *Analyser) {
	if p, ok := node.(*phrase.Phrase); ok {
		analyser.Preorder(p)

		for _, child := range p.Children {
			Traverse(child, analyser)
		}

		analyser.Postorder(p)
	}
}
