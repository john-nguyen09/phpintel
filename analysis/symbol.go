package analysis

import (
	"github.com/john-nguyen09/phpintel/analysis/storage"
	"github.com/john-nguyen09/phpintel/internal/lsp/protocol"
	sitter "github.com/smacker/go-tree-sitter"
)

// VisibilityModifierValue is a value of visibility modifier (public, protected, private)
type VisibilityModifierValue int

const (
	// Public indicates public visibility
	Public VisibilityModifierValue = iota
	// Protected indicates protected visibility
	Protected = iota
	// Private indicates private visibility
	Private = iota
)

func (v VisibilityModifierValue) ToString() string {
	if v == Public {
		return "public"
	}
	if v == Private {
		return "private"
	}
	if v == Protected {
		return "protected"
	}
	return ""
}

// ClassModifierValue is a value of class modifier (abstract, final)
type ClassModifierValue int

const (
	// NoClassModifier indicates there is no class modifier
	NoClassModifier ClassModifierValue = iota
	// Abstract indicates there is an abstract keyword
	Abstract = iota
	// Final indicates there is a final keyword
	Final = iota
)

// Symbol is a symbol
type Symbol interface {
	GetLocation() protocol.Location
}

// NameIndexable indicates a symbol is name indexable, i.e. have completion
type NameIndexable interface {
	GetIndexableName() string
	GetIndexCollection() string
}

type HasParams interface {
	GetParams() []*Parameter
	GetDescription() string
	GetNameLabel() string
}

type HasScope interface {
	GetScope() string
	IsScopeSymbol() bool
}

type serialisable interface {
	GetCollection() string
	GetKey() string
	Serialise(*storage.Encoder)
}

type HasParamsResolvable interface {
	ResolveToHasParams(ctx ResolveContext) []HasParams
}

func transformQualifiedName(n *sitter.Node, document *Document) TypeString {
	return NewTypeString(document.GetNodeText(n))
}
