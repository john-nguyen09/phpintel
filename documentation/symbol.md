# Symbol
## Definition
- Is internal structure for indexer to understand
- Can have multiple types:
    - Symbol
    - TransformSymbol
    - CollectionSymbol
    - TokenSymbol
- Interfaces:
    - Consumer
    - Reference
    - ScopeMember
- Is indexed when File consumes