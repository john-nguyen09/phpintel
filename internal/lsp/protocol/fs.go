package protocol

import (
	"context"
	"io/ioutil"

	"github.com/john-nguyen09/phpintel/internal/jsonrpc2"
	"github.com/john-nguyen09/phpintel/util"
	"github.com/karrick/godirwalk"
)

// FS is an interface for reading files
type FS interface {
	ConvertToURI(uri string) string
	ReadFile(ctx context.Context, uri string) ([]byte, error)
	ListFiles(ctx context.Context, base string) ([]TextDocumentIdentifier, error)
}

// URIFS is a LSP file reader which relies on xcontentProvider and xfilesProvider
// client capabilities
type URIFS struct {
	conn *jsonrpc2.Conn
}

var _ FS = (*URIFS)(nil)

// NewLSPFS creates an URIFS instance from the connection
func NewLSPFS(conn *jsonrpc2.Conn) *URIFS {
	return &URIFS{
		conn: conn,
	}
}

// ConvertToURI does nothing and return the path because it's already URI
func (f *URIFS) ConvertToURI(path string) string {
	return path
}

// ReadFile uses textDocument/xcontent request to read content of the uri
func (f *URIFS) ReadFile(ctx context.Context, uri string) ([]byte, error) {
	var textDocument TextDocumentItem
	err := f.conn.Call(ctx, "textDocument/xcontent", &ContentParams{
		TextDocument: TextDocumentIdentifier{
			URI: uri,
		},
	}, &textDocument)
	if err != nil {
		return nil, err
	}
	return []byte(textDocument.Text), nil
}

// ListFiles uses workspace/xfiles request to list files under a base
func (f *URIFS) ListFiles(ctx context.Context, base string) (docs []TextDocumentIdentifier, err error) {
	err = f.conn.Call(ctx, "workspace/xfiles", FilesParams{Base: base}, &docs)
	return
}

// FileFS uses file path
type FileFS struct{}

// NewFileFS creates a FileFS instance
func NewFileFS() *FileFS {
	return &FileFS{}
}

var _ FS = (*FileFS)(nil)

// ConvertToURI converts path to URI
func (f *FileFS) ConvertToURI(path string) string {
	return util.PathToURI(path)
}

// ReadFile reads the file from uri, given that the uri is the file path
// rather than actually the URI
func (f *FileFS) ReadFile(ctx context.Context, uri string) ([]byte, error) {
	path, err := util.URIToPath(uri)
	if err != nil {
		return nil, err
	}
	return ioutil.ReadFile(path)
}

// ListFiles lists files from the base
func (f *FileFS) ListFiles(ctx context.Context, base string) ([]TextDocumentIdentifier, error) {
	var results []TextDocumentIdentifier
	godirwalk.Walk(base, &godirwalk.Options{
		Callback: func(path string, de *godirwalk.Dirent) error {
			if !de.IsDir() {
				results = append(results, TextDocumentIdentifier{
					URI: path,
				})
			}
			return nil
		},
		Unsorted: true,
	})
	return results, nil
}
