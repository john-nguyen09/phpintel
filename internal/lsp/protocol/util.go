package protocol

import (
	"strconv"
	"strings"
)

// IsInRange compare whether a position is within a range
func IsInRange(position Position, theRange Range) int {
	start := theRange.Start
	end := theRange.End

	if position.Line < start.Line || (position.Line == start.Line && position.Character < start.Character) {
		return -1
	}
	if position.Line > end.Line || (position.Line == end.Line && position.Character >= end.Character) {
		if ComparePos(start, end) == 0 {
			return -1
		}
		return 1
	}
	return 0
}

// ComparePos returns -1, 0 or 1 indicating whether a is before, equal to or after b
func ComparePos(a Position, b Position) int {
	if a.Line < b.Line {
		return -1
	}
	if a.Line == b.Line {
		if a.Character < b.Character {
			return -1
		}
		if a.Character == b.Character {
			return 0
		}
		if a.Character > b.Character {
			return 1
		}
	}
	return 1
}

// CompareRange compares 2 ranges
func CompareRange(a Range, b Range) int {
	if IsInRange(a.End, b) < 0 {
		return -1
	}
	if IsInRange(a.Start, b) > 0 {
		return 1
	}
	return 0
}

// PosFromString creates Position from string
func PosFromString(str string) Position {
	values := strings.Split(str, ":")
	line, err := strconv.Atoi(values[0])
	if err != nil {
		panic(err)
	}
	character, err := strconv.Atoi(values[1])
	if err != nil {
		panic(err)
	}
	return Position{
		Line:      line,
		Character: character,
	}
}

// RangeFromString creates Range from string
func RangeFromString(str string) Range {
	values := strings.Split(str, "-")
	start := PosFromString(values[0])
	end := PosFromString(values[1])
	return Range{
		Start: start,
		End:   end,
	}
}
