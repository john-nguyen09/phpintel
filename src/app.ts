import "reflect-metadata";
import { LevelDatasource } from "./storage/db";
import { LogWriter } from "./service/logWriter";
import { Hasher } from "./service/hasher";
import { Container, interfaces } from "inversify";
import { Indexer } from "./index/indexer";
import { ClassTable } from "./storage/table/class";
import { ClassConstantTable } from "./storage/table/classConstant";
import { ConstantTable } from "./storage/table/constant";
import { FunctionTable } from "./storage/table/function";
import { MethodTable } from "./storage/table/method";
import { PropertyTable } from "./storage/table/property";
import { Traverser } from "./traverser";
import { ReferenceTable } from "./storage/table/reference";
import { PhpDocumentTable } from "./storage/table/phpDoc";
import { IConnection } from "vscode-languageserver";
import { ScopeVarTable } from "./storage/table/scopeVar";

export interface AppOptions {
    storage: string;
}

export class Application {
    protected container: Container = new Container();
    protected options: AppOptions = {
        storage: ''
    };

    constructor(storageLocation: string, connection?: IConnection) {
        this.container.bind<IConnection | undefined>('IConnection').toConstantValue(connection);

        this.initBind();
        this.initStorage(storageLocation);
    }

    public getContainer(): Container {
        return this.container;
    }

    public async clearCache() {
        const db = this.container.get<LevelDatasource>(LevelDatasource).getDb();
        let promises: Promise<void>[] = [];

        return new Promise<void>((resolve, reject) => {
            db.createKeyStream()
                .on('data', (key) => {
                    promises.push(db.del(key));
                })
                .on('end', () => {
                    Promise.all(promises).then(() => { resolve() });
                })
                .on('error', (err: Error) => {
                    reject(err);
                });
        });
    }

    public async shutdown() {
        await this.container.get<LevelDatasource>(LevelDatasource).getDb().close();
    }

    protected initStorage(location: string) {
        this.options.storage = location;

        let datasource = new LevelDatasource(location);
        this.container.bind<LevelDatasource>(LevelDatasource).toConstantValue(datasource);

        // Tables
        this.container.bind<ClassTable>(ClassTable).toSelf().inSingletonScope();
        this.container.bind<ClassConstantTable>(ClassConstantTable).toSelf().inSingletonScope();
        this.container.bind<ConstantTable>(ConstantTable).toSelf().inSingletonScope();
        this.container.bind<FunctionTable>(FunctionTable).toSelf().inSingletonScope();
        this.container.bind<MethodTable>(MethodTable).toSelf().inSingletonScope();
        this.container.bind<PropertyTable>(PropertyTable).toSelf().inSingletonScope();
        this.container.bind<PhpDocumentTable>(PhpDocumentTable).toSelf().inSingletonScope();
        this.container.bind<ReferenceTable>(ReferenceTable).toSelf().inSingletonScope();
        this.container.bind<ScopeVarTable>(ScopeVarTable).toSelf().inSingletonScope();
    }

    protected initBind() {
        this.container.bind<LogWriter>(LogWriter).toSelf().inSingletonScope();
        this.container.bind<Hasher>(Hasher).toSelf().inSingletonScope();
        this.container.bind<Indexer>(Indexer).toSelf().inSingletonScope();
        this.container.bind<Traverser>(Traverser).toSelf().inSingletonScope();
    }
}

export namespace App {
    let app: Application;

    export function init(storageLocation: string, connection?: IConnection) {
        app = new Application(storageLocation, connection);
    }

    export function get<T>(
        identifier: string | symbol | interfaces.Newable<T> | interfaces.Abstract<T>
    ): T {
        return app.getContainer().get(identifier);
    }

    export function set<T>(
        identifier: string,
        value: T
    ): void {
        app.getContainer().bind<T>(identifier).toConstantValue(value);
    }

    export async function clearCache() {
        return await app.clearCache();
    }

    export async function shutdown() {
        return await app.shutdown();
    }
}
