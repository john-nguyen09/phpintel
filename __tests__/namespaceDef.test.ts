import "reflect-metadata";
import { SymbolParser } from "../src/symbol/symbolParser";
import { PhpDocument } from "../src/symbol/phpDocument";
import { pathToUri } from "../src/util/uri";
import * as path from 'path';
import * as fs from 'fs';
import { Parser } from "php7parser";
import { Traverser } from "../src/traverser";
import { App } from "../src/app";
import { getDebugDir, getCaseDir } from "../src/testHelper";
import { Indexer, PhpFileInfo } from "../src/index/indexer";
import { PhpDocumentTable } from "../src/storage/table/phpDoc";

beforeAll(() => {
    App.init(path.join(getDebugDir(), 'storage'));
});

beforeEach(async () => {
    await App.clearCache();
});

afterAll(async () => {
    await App.shutdown();
});

describe('namespaceDef', () => {
    it('should assign namespace to phpDocument', () => {
        let workspaceDir = path.resolve(__dirname, '..', 'case', 'namespaceDef');
        let files = fs.readdirSync(workspaceDir);
        let treeTraverser = new Traverser();

        for (let file of files) {
            let filePath = path.join(workspaceDir, file);
            let fileUri = pathToUri(filePath);

            if (file.endsWith('.php')) {
                const fileContent = fs.readFileSync(filePath).toString();
                let symbolParser = new SymbolParser(new PhpDocument(
                    fileUri,
                    fileContent
                ));
                let parseTree = Parser.parse(fileContent);

                treeTraverser.traverse(parseTree, [
                    symbolParser
                ]);

                expect(symbolParser.getPhpDoc().toObject()).toMatchSnapshot();
            }
        }
    });

    it('returns import table', async () => {
        const indexer = App.get<Indexer>(Indexer);
        const phpDocTable = App.get<PhpDocumentTable>(PhpDocumentTable);

        const filePath = path.join(getCaseDir(), 'namespaceDef', 'import_table.php');
        const fileUri = pathToUri(filePath);

        await indexer.syncFileSystem(await PhpFileInfo.createFileInfo(filePath));

        let phpDoc = await phpDocTable.get(fileUri);

        if (phpDoc === null) {
            return;
        }

        expect(phpDoc.importTable).toMatchSnapshot();
    });
});