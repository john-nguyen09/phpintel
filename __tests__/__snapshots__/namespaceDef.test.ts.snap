// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`namespaceDef returns import table 1`] = `
ImportTable {
  "imports": Object {
    "": "\\\\NamespaceDef\\\\TestClassNamespace1",
  },
  "namespace": NamespaceName {
    "parts": Array [
      "ImportTable",
    ],
  },
}
`;

exports[`namespaceDef should assign namespace to phpDocument 1`] = `
PhpDocument {
  "argumentLists": Array [],
  "classConstants": Array [],
  "classes": Array [
    Class {
      "extend": null,
      "implements": Array [],
      "location": Object {
        "range": Object {
          "end": 61,
          "start": 31,
        },
        "uri": "case/namespaceDef/class.php",
      },
      "modifier": SymbolModifier {
        "modifier": 0,
        "visibility": 0,
      },
      "name": TypeName {
        "name": "\\\\NamespaceDefTestClassNamespace1",
      },
      "traits": Array [],
    },
  ],
  "constants": Array [],
  "functions": Array [],
  "importTable": ImportTable {
    "imports": Object {},
    "namespace": NamespaceName {
      "parts": Array [
        "NamespaceDef",
      ],
    },
  },
  "methods": Array [],
  "modifiedTime": -1,
  "properties": Array [],
  "references": Array [],
  "scopeVarStack": Array [
    ScopeVar {
      "location": Object {
        "range": Object {
          "end": 61,
          "start": 0,
        },
        "uri": "case/namespaceDef/class.php",
      },
      "variables": Object {},
    },
  ],
  "text": "<?php
namespace NamespaceDef;

class TestClassNamespace1 {

}",
  "uri": "case/namespaceDef/class.php",
}
`;

exports[`namespaceDef should assign namespace to phpDocument 2`] = `
PhpDocument {
  "argumentLists": Array [],
  "classConstants": Array [],
  "classes": Array [],
  "constants": Array [
    Constant {
      "name": TypeName {
        "name": "\\\\NamespaceDefPI",
      },
      "type": TypeName {
        "name": "float",
      },
      "value": "3.14",
    },
  ],
  "functions": Array [],
  "importTable": ImportTable {
    "imports": Object {},
    "namespace": NamespaceName {
      "parts": Array [
        "NamespaceDef",
      ],
    },
  },
  "methods": Array [],
  "modifiedTime": -1,
  "properties": Array [],
  "references": Array [
    Constant {
      "name": TypeName {
        "name": "\\\\NamespaceDefPI",
      },
      "type": TypeName {
        "name": "float",
      },
      "value": "3.14",
    },
  ],
  "scopeVarStack": Array [
    ScopeVar {
      "location": Object {
        "range": Object {
          "end": 47,
          "start": 0,
        },
        "uri": "case/namespaceDef/constant.php",
      },
      "variables": Object {},
    },
  ],
  "text": "<?php
namespace NamespaceDef;

const PI = 3.14;",
  "uri": "case/namespaceDef/constant.php",
}
`;

exports[`namespaceDef should assign namespace to phpDocument 3`] = `
PhpDocument {
  "argumentLists": Array [
    ArgumentExpressionList {
      "arguments": Array [],
      "location": Object {
        "range": Object {
          "end": 70,
          "start": 38,
        },
        "uri": "case/namespaceDef/defineConst.php",
      },
    },
  ],
  "classConstants": Array [],
  "classes": Array [],
  "constants": Array [
    DefineConstant {
      "name": TypeName {
        "name": "\\\\NamespaceDefDEFINE_CONST",
      },
      "type": TypeName {
        "name": "string",
      },
      "value": "'Hello world!!!'",
    },
  ],
  "functions": Array [],
  "importTable": ImportTable {
    "imports": Object {},
    "namespace": NamespaceName {
      "parts": Array [
        "NamespaceDef",
      ],
    },
  },
  "methods": Array [],
  "modifiedTime": -1,
  "properties": Array [],
  "references": Array [
    FunctionCall {
      "argumentList": ArgumentExpressionList {
        "arguments": Array [],
        "location": Object {
          "range": Object {
            "end": 70,
            "start": 38,
          },
          "uri": "case/namespaceDef/defineConst.php",
        },
      },
      "location": Object {
        "range": Object {
          "end": 71,
          "start": 31,
        },
        "uri": "case/namespaceDef/defineConst.php",
      },
      "realSymbol": DefineConstant {
        "name": TypeName {
          "name": "\\\\NamespaceDefDEFINE_CONST",
        },
        "type": TypeName {
          "name": "string",
        },
        "value": "'Hello world!!!'",
      },
      "refKind": 1,
      "scope": null,
      "type": TypeName {
        "name": "",
      },
    },
    DefineConstant {
      "name": TypeName {
        "name": "\\\\NamespaceDefDEFINE_CONST",
      },
      "type": TypeName {
        "name": "string",
      },
      "value": "'Hello world!!!'",
    },
  ],
  "scopeVarStack": Array [
    ScopeVar {
      "location": Object {
        "range": Object {
          "end": 72,
          "start": 0,
        },
        "uri": "case/namespaceDef/defineConst.php",
      },
      "variables": Object {},
    },
  ],
  "text": "<?php
namespace NamespaceDef;

define('DEFINE_CONST', 'Hello world!!!');",
  "uri": "case/namespaceDef/defineConst.php",
}
`;

exports[`namespaceDef should assign namespace to phpDocument 4`] = `
PhpDocument {
  "argumentLists": Array [],
  "classConstants": Array [],
  "classes": Array [],
  "constants": Array [],
  "functions": Array [
    Function {
      "description": "",
      "name": TypeName {
        "name": "\\\\NamespaceDeffunction_with_namespace",
      },
      "parameters": Array [
        Parameter {
          "description": "",
          "location": Object {
            "range": Object {
              "end": 215,
              "start": 208,
            },
            "uri": "case/namespaceDef/function.php",
          },
          "name": "$param1",
          "refKind": 3,
          "scope": null,
          "type": TypeComposite {
            "_types": Array [
              TypeName {
                "name": "\\\\NamespaceDef\\\\TestClassNamespace1",
              },
            ],
            "existingTypes": Object {
              "TestClassNamespace1": true,
            },
          },
          "value": "",
        },
        Parameter {
          "description": "",
          "location": Object {
            "range": Object {
              "end": 224,
              "start": 217,
            },
            "uri": "case/namespaceDef/function.php",
          },
          "name": "$param2",
          "refKind": 3,
          "scope": null,
          "type": TypeComposite {
            "_types": Array [
              TypeName {
                "name": "\\\\FullyQualified\\\\Name\\\\Class",
              },
            ],
            "existingTypes": Object {
              "\\\\FullyQualified\\\\Name\\\\Class": true,
            },
          },
          "value": "",
        },
        Parameter {
          "description": "",
          "location": Object {
            "range": Object {
              "end": 238,
              "start": 226,
            },
            "uri": "case/namespaceDef/function.php",
          },
          "name": "$param3",
          "refKind": 3,
          "scope": null,
          "type": TypeComposite {
            "_types": Array [
              TypeName {
                "name": "boolean",
              },
            ],
            "existingTypes": Object {
              "boolean": true,
            },
          },
          "value": "",
        },
      ],
      "scopeVar": ScopeVar {
        "location": Object {
          "range": Object {
            "end": 579,
            "start": 175,
          },
          "uri": "case/namespaceDef/function.php",
        },
        "variables": Object {
          "$param1": TypeComposite {
            "_types": Array [
              TypeName {
                "name": "\\\\NamespaceDef\\\\TestClassNamespace1",
              },
            ],
            "existingTypes": Object {
              "TestClassNamespace1": true,
            },
          },
          "$param2": TypeComposite {
            "_types": Array [
              TypeName {
                "name": "\\\\FullyQualified\\\\Name\\\\Class",
              },
            ],
            "existingTypes": Object {
              "\\\\FullyQualified\\\\Name\\\\Class": true,
            },
          },
          "$param3": TypeComposite {
            "_types": Array [
              TypeName {
                "name": "boolean",
              },
            ],
            "existingTypes": Object {
              "boolean": true,
            },
          },
          "$var1": TypeComposite {
            "_types": Array [
              TypeName {
                "name": "boolean",
              },
            ],
            "existingTypes": Object {
              "boolean": true,
            },
          },
          "$var2": TypeComposite {
            "_types": Array [
              TypeName {
                "name": "string",
              },
            ],
            "existingTypes": Object {
              "string": true,
            },
          },
          "$var3": TypeComposite {
            "_types": Array [
              TypeName {
                "name": "\\\\NamespaceDef\\\\CONSTANT_IN_NAMESPACE",
              },
            ],
            "existingTypes": Object {
              "CONSTANT_IN_NAMESPACE": true,
            },
          },
        },
      },
      "types": Array [
        TypeName {
          "name": "\\\\NamespaceDef\\\\CONSTANT_IN_NAMESPACE",
        },
        TypeName {
          "name": "string",
        },
        TypeName {
          "name": "\\\\NamespaceDef\\\\TestClassNamespace1",
        },
        TypeName {
          "name": "\\\\FullyQualified\\\\Name\\\\Class",
        },
        TypeName {
          "name": "boolean",
        },
      ],
    },
  ],
  "importTable": ImportTable {
    "imports": Object {},
    "namespace": NamespaceName {
      "parts": Array [
        "NamespaceDef",
      ],
    },
  },
  "methods": Array [],
  "modifiedTime": -1,
  "properties": Array [],
  "references": Array [
    Parameter {
      "description": "",
      "location": Object {
        "range": Object {
          "end": 215,
          "start": 208,
        },
        "uri": "case/namespaceDef/function.php",
      },
      "name": "$param1",
      "refKind": 3,
      "scope": null,
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "\\\\NamespaceDef\\\\TestClassNamespace1",
          },
        ],
        "existingTypes": Object {
          "TestClassNamespace1": true,
        },
      },
      "value": "",
    },
    Parameter {
      "description": "",
      "location": Object {
        "range": Object {
          "end": 224,
          "start": 217,
        },
        "uri": "case/namespaceDef/function.php",
      },
      "name": "$param2",
      "refKind": 3,
      "scope": null,
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "\\\\FullyQualified\\\\Name\\\\Class",
          },
        ],
        "existingTypes": Object {
          "\\\\FullyQualified\\\\Name\\\\Class": true,
        },
      },
      "value": "",
    },
    Parameter {
      "description": "",
      "location": Object {
        "range": Object {
          "end": 238,
          "start": 226,
        },
        "uri": "case/namespaceDef/function.php",
      },
      "name": "$param3",
      "refKind": 3,
      "scope": null,
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "boolean",
          },
        ],
        "existingTypes": Object {
          "boolean": true,
        },
      },
      "value": "",
    },
    SimpleVariable {
      "name": "$var1",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "boolean",
          },
        ],
        "existingTypes": Object {
          "boolean": true,
        },
      },
    },
    ConstantAccess {
      "location": Object {
        "range": Object {
          "end": 259,
          "start": 254,
        },
        "uri": "case/namespaceDef/function.php",
      },
      "refKind": 6,
      "scope": null,
      "type": TypeName {
        "name": "boolean",
      },
      "value": "false",
    },
    SimpleVariable {
      "name": "$var2",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "string",
          },
        ],
        "existingTypes": Object {
          "string": true,
        },
      },
    },
    SimpleVariable {
      "name": "$var3",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "\\\\NamespaceDef\\\\CONSTANT_IN_NAMESPACE",
          },
        ],
        "existingTypes": Object {
          "CONSTANT_IN_NAMESPACE": true,
        },
      },
    },
    ConstantAccess {
      "location": Object {
        "range": Object {
          "end": 313,
          "start": 292,
        },
        "uri": "case/namespaceDef/function.php",
      },
      "refKind": 6,
      "scope": null,
      "type": TypeName {
        "name": "\\\\NamespaceDef\\\\CONSTANT_IN_NAMESPACE",
      },
      "value": "CONSTANT_IN_NAMESPACE",
    },
    SimpleVariable {
      "name": "$var1",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "boolean",
          },
        ],
        "existingTypes": Object {
          "boolean": true,
        },
      },
    },
    SimpleVariable {
      "name": "$var2",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "string",
          },
        ],
        "existingTypes": Object {
          "string": true,
        },
      },
    },
    SimpleVariable {
      "name": "$var3",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "\\\\NamespaceDef\\\\CONSTANT_IN_NAMESPACE",
          },
        ],
        "existingTypes": Object {
          "CONSTANT_IN_NAMESPACE": true,
        },
      },
    },
    SimpleVariable {
      "name": "$var3",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "\\\\NamespaceDef\\\\CONSTANT_IN_NAMESPACE",
          },
        ],
        "existingTypes": Object {
          "CONSTANT_IN_NAMESPACE": true,
        },
      },
    },
    SimpleVariable {
      "name": "$var1",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "boolean",
          },
        ],
        "existingTypes": Object {
          "boolean": true,
        },
      },
    },
    SimpleVariable {
      "name": "$var2",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "string",
          },
        ],
        "existingTypes": Object {
          "string": true,
        },
      },
    },
    SimpleVariable {
      "name": "$var2",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "string",
          },
        ],
        "existingTypes": Object {
          "string": true,
        },
      },
    },
    SimpleVariable {
      "name": "$var1",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "boolean",
          },
        ],
        "existingTypes": Object {
          "boolean": true,
        },
      },
    },
    SimpleVariable {
      "name": "$param1",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "\\\\NamespaceDef\\\\TestClassNamespace1",
          },
        ],
        "existingTypes": Object {
          "TestClassNamespace1": true,
        },
      },
    },
    SimpleVariable {
      "name": "$var2",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "string",
          },
        ],
        "existingTypes": Object {
          "string": true,
        },
      },
    },
    SimpleVariable {
      "name": "$param2",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "\\\\FullyQualified\\\\Name\\\\Class",
          },
        ],
        "existingTypes": Object {
          "\\\\FullyQualified\\\\Name\\\\Class": true,
        },
      },
    },
    SimpleVariable {
      "name": "$var3",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "\\\\NamespaceDef\\\\CONSTANT_IN_NAMESPACE",
          },
        ],
        "existingTypes": Object {
          "CONSTANT_IN_NAMESPACE": true,
        },
      },
    },
    SimpleVariable {
      "name": "$param3",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "boolean",
          },
        ],
        "existingTypes": Object {
          "boolean": true,
        },
      },
    },
  ],
  "scopeVarStack": Array [
    ScopeVar {
      "location": Object {
        "range": Object {
          "end": 579,
          "start": 0,
        },
        "uri": "case/namespaceDef/function.php",
      },
      "variables": Object {},
    },
    ScopeVar {
      "location": Object {
        "range": Object {
          "end": 579,
          "start": 175,
        },
        "uri": "case/namespaceDef/function.php",
      },
      "variables": Object {
        "$param1": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "\\\\NamespaceDef\\\\TestClassNamespace1",
            },
          ],
          "existingTypes": Object {
            "TestClassNamespace1": true,
          },
        },
        "$param2": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "\\\\FullyQualified\\\\Name\\\\Class",
            },
          ],
          "existingTypes": Object {
            "\\\\FullyQualified\\\\Name\\\\Class": true,
          },
        },
        "$param3": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "boolean",
            },
          ],
          "existingTypes": Object {
            "boolean": true,
          },
        },
        "$var1": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "boolean",
            },
          ],
          "existingTypes": Object {
            "boolean": true,
          },
        },
        "$var2": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "string",
            },
          ],
          "existingTypes": Object {
            "string": true,
          },
        },
        "$var3": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "\\\\NamespaceDef\\\\CONSTANT_IN_NAMESPACE",
            },
          ],
          "existingTypes": Object {
            "CONSTANT_IN_NAMESPACE": true,
          },
        },
      },
    },
  ],
  "text": "<?php
namespace NamespaceDef;

/**
 * @param TestClassNamespace1 $param1
 * @param \\\\FullyQualified\\\\Name\\\\Class $param2
 * @param boolean
 * 
 * @return TestClassNamespace2
 */
function function_with_namespace($param1, $param2, bool $param3) {
    $var1 = false;
    $var2 = 'abc';
    $var3 = CONSTANT_IN_NAMESPACE;

    if ($var1 && $var2 && $var3) {
        return $var3;
    }

    if ($var1 && $var2) {
        return $var2;
    }

    if ($var1) {
        return $param1;
    }

    if ($var2) {
        return $param2;
    }

    if ($var3) {
        return $param3;
    }
}",
  "uri": "case/namespaceDef/function.php",
}
`;

exports[`namespaceDef should assign namespace to phpDocument 5`] = `
PhpDocument {
  "argumentLists": Array [],
  "classConstants": Array [],
  "classes": Array [],
  "constants": Array [],
  "functions": Array [],
  "importTable": ImportTable {
    "imports": Object {
      "": "\\\\NamespaceDef\\\\TestClassNamespace1",
    },
    "namespace": NamespaceName {
      "parts": Array [
        "ImportTable",
      ],
    },
  },
  "methods": Array [],
  "modifiedTime": -1,
  "properties": Array [],
  "references": Array [],
  "scopeVarStack": Array [
    ScopeVar {
      "location": Object {
        "range": Object {
          "end": 149,
          "start": 0,
        },
        "uri": "case/namespaceDef/import_table.php",
      },
      "variables": Object {},
    },
  ],
  "text": "<?php
namespace ImportTable;

use NamespaceDef\\\\{
    const PI,
    function function_with_namespace
};
use NamespaceDef\\\\TestClassNamespace1;
",
  "uri": "case/namespaceDef/import_table.php",
}
`;
